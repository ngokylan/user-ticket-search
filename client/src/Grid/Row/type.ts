import { ReactNode } from 'react';

export type RowDirection =
  | null
  | 'row'
  | 'column'
  | 'row-reverse'
  | 'column-reverse';
export type RowWrap = null | 'wrap' | 'nowrap' | 'wrap-reverse';
export type RowJustifyContent =
  | null
  | 'start'
  | 'end'
  | 'center'
  | 'between'
  | 'around';
export type RowAlignItems =
  | null
  | 'start'
  | 'end'
  | 'center'
  | 'baseline'
  | 'stretch';
export type RowAlignContent =
  | null
  | 'start'
  | 'end'
  | 'center'
  | 'between'
  | 'around'
  | 'stretch';

export type RowProps = {
  /** Content of row */
  children?: ReactNode;
  /** order atribute of row */
  isNoGutters?: boolean;
  /** order atribute of row */
  isInline?: boolean;
  /**
   * order atribute of row, use one from the following:
   * RowDirection = null | 'row' | 'column' | 'row-reverse' | 'column-reverse'
   */
  direction?: RowDirection;
  /**
   * order atribute of row, use one from the following:
   * RowWrap = null | 'wrap' | 'nowrap' | 'wrap-reverse'
   */
  wrap?: RowWrap;
  /**
   * order atribute of row, use one from the following:
   * RowJustifyContent = null | 'start' | 'end' | 'center' | 'between' | 'around'
   */
  justifyContent?: RowJustifyContent;
  /**
   * order atribute of row, use one from the following:
   * RowAlignItems = null | 'start' | 'end' | 'center' | 'baseline' | 'stretch'
   */
  alignItems?: RowAlignItems;
  /**
   * order atribute of row, use one from the following:
   * RowAlignContent = null | 'start' | 'end' | 'center' | 'between' | 'around' | 'stretch'
   */
  alignContent?: RowAlignContent;
  /** xs responsive atribute of row */
  xs?: object;
  /** sm responsive atribute of row */
  sm?: object;
  /** md responsive atribute of row */
  md?: object;
  /** lg responsive atribute of row */
  lg?: object;
  /** xl responsive atribute of row */
  xl?: object;
  /** xxl responsive atribute of row */
  xxl?: object;
  /** key atribute of col */
  key?: string | number;
  /** Row class name attribute, (optional) */
  className?: string;
  /** WCAG role */
  role?: string;
};

export type RowResponsivePropertiesType = {
  property: string;
  prefix: string;
};
