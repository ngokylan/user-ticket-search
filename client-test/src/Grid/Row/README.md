
No Gutters
```js
import Col from '../Col';
import Container from '../Container';

<Container>
  <Row isNoGutters isStyleguidist>
    <Col className="demo-col">A</Col>
    <Col className="demo-col">B</Col>
    <Col className="demo-col">C</Col>
  </Row>
</Container>
```

Inline
```js
import Col from '../Col';
import Container from '../Container';

<Container>
  <Row isStyleguidist isInline>
    <Col className="demo-col">A</Col>
    <Col className="demo-col">B</Col>
    <Col className="demo-col">C</Col>
  </Row>
</Container>
```

Direction
```js
import Col from '../Col';
import Container from '../Container';

<Container>
  <Row isStyleguidist direction="row">
    <Col className="demo-col">A</Col>
    <Col className="demo-col">B</Col>
    <Col className="demo-col">C</Col>
  </Row>
<br/>
  <Row isStyleguidist direction="column">
    <Col className="demo-col" basis="">A</Col>
    <Col className="demo-col">B</Col>
    <Col className="demo-col">C</Col>
  </Row>
<br/>
  <Row isStyleguidist direction="row-reverse">
    <Col className="demo-col">A</Col>
    <Col className="demo-col">B</Col>
    <Col className="demo-col">C</Col>
  </Row>
<br/>
  <Row isStyleguidist isInline direction="column-reverse">
    <Col className="demo-col">A</Col>
    <Col className="demo-col">B</Col>
    <Col className="demo-col">C</Col>
  </Row>
</Container>
```

Wrap
```js
import Col from '../Col';
import Container from '../Container';

<Container>
  <Row isStyleguidist wrap="wrap">
    <Col className="demo-col" span={12}>A</Col>
    <Col className="demo-col" span={12}>B</Col>
    <Col className="demo-col" span={12}>C</Col>
  </Row>
<br/>
<div style={{maxWidth:'100%',overflow:'auto'}}>
  <Row isStyleguidist wrap="nowrap">
    <Col className="demo-col" span={12}>A</Col>
    <Col className="demo-col" span={12}>B</Col>
    <Col className="demo-col" span={12}>C</Col>
  </Row>
</div>
<br/>
  <Row isStyleguidist wrap="wrap-reverse">
    <Col className="demo-col" span={12}>A</Col>
    <Col className="demo-col" span={12}>B</Col>
    <Col className="demo-col" span={12}>C</Col>
  </Row>
</Container>
```

Justify Content
```js
import Col from '../Col';
import Container from '../Container';

<Container>
  <Row isStyleguidist justifyContent="start">
    <Col className="demo-col" span={2}>A</Col>
    <Col className="demo-col" span={2}>B</Col>
    <Col className="demo-col" span={2}>C</Col>
  </Row>
<br/>
  <Row isStyleguidist justifyContent="end">
    <Col className="demo-col" span={2}>A</Col>
    <Col className="demo-col" span={2}>B</Col>
    <Col className="demo-col" span={2}>C</Col>
  </Row>
<br/>
  <Row isStyleguidist justifyContent="center">
    <Col className="demo-col" span={2}>A</Col>
    <Col className="demo-col" span={2}>B</Col>
    <Col className="demo-col" span={2}>C</Col>
  </Row>
<br/>
  <Row isStyleguidist justifyContent="between">
    <Col className="demo-col" span={2}>A</Col>
    <Col className="demo-col" span={2}>B</Col>
    <Col className="demo-col" span={2}>C</Col>
  </Row>
<br/>
  <Row isStyleguidist justifyContent="around">
    <Col className="demo-col" span={2}>A</Col>
    <Col className="demo-col" span={2}>B</Col>
    <Col className="demo-col" span={2}>C</Col>
  </Row>
<br/>
</Container>
```

Align Items
```js
import Col from '../Col';
import Container from '../Container';

<Container>
  <Row isStyleguidist alignItems="start">
    <Col className="demo-col" span={2}>A</Col>
    <Col className="demo-col" span={2}><br />B<br /><br /></Col>
    <Col className="demo-col" span={2}>C</Col>
  </Row>
<br />
  <Row isStyleguidist alignItems="end">
    <Col className="demo-col" span={2}>A</Col>
    <Col className="demo-col" span={2}><br />B<br /><br /></Col>
    <Col className="demo-col" span={2}>C</Col>
  </Row>
<br />
  <Row isStyleguidist alignItems="center">
    <Col className="demo-col" span={2}>A</Col>
    <Col className="demo-col" span={2}><br />B<br /><br /></Col>
    <Col className="demo-col" span={2}>C</Col>
  </Row>
<br />
  <Row isStyleguidist alignItems="baseline">
    <Col className="demo-col" span={2}>A</Col>
    <Col className="demo-col" span={2}><br />B<br /><br /></Col>
    <Col className="demo-col" span={2}>C</Col>
  </Row>
<br />
  <Row isStyleguidist alignItems="stretch">
    <Col className="demo-col" span={2}>A</Col>
    <Col className="demo-col" span={2}><br />B<br /><br /></Col>
    <Col className="demo-col" span={2}>C</Col>
  </Row>
</Container>
```

Align Content
```js
import Col from '../Col';
import Container from '../Container';

<Container>
  <Row alignItems="stretch">
    <Col>
      <Row alignContent="start">
        <Col className="demo-col" span={12}>A</Col>
        <Col className="demo-col" span={12}><br />B<br /><br /></Col>
        <Col className="demo-col" span={12}>C</Col>
      </Row>
    </Col>
    <Col>
      <Row alignContent="end">
        <Col className="demo-col" span={12}>A</Col>
        <Col className="demo-col" span={12}><br />B<br /><br /></Col>
        <Col className="demo-col" span={12}>C</Col>
      </Row>
    </Col>
    <Col>
      <Row alignContent="center">
        <Col className="demo-col" span={12}>A</Col>
        <Col className="demo-col" span={12}><br />B<br /><br /></Col>
        <Col className="demo-col" span={12}>C</Col>
      </Row>
    </Col>
    <Col>
      <Row alignContent="between">
        <Col className="demo-col" span={12}>A</Col>
        <Col className="demo-col" span={12}><br />B<br /><br /></Col>
        <Col className="demo-col" span={12}>C</Col>
      </Row>
    </Col>
    <Col>
      <Row alignContent="around">
        <Col className="demo-col" span={12}>A</Col>
        <Col className="demo-col" span={12}><br />B<br /><br /></Col>
        <Col className="demo-col" span={12}>C</Col>
      </Row>
    </Col>
    <Col>
      <Row alignContent="stretch">
        <Col className="demo-col" span={12}>A</Col>
        <Col className="demo-col" span={12}><br />B<br /><br /></Col>
        <Col className="demo-col" span={12}>C</Col>
      </Row>
    </Col>
    <Col><div style={{height:'200px'}}></div></Col>
  </Row>
</Container>
```

The props above also can be used in specific breakpoint.
```js
import Col from '../Col';
import Container from '../Container';

<Container>
  <Row isInline sm={{display:'block',direction:'row-reverse'}} lg={{direction:'row'}}>
    <Col className="demo-col">A</Col>
    <Col className="demo-col">B</Col>
    <Col className="demo-col">C</Col>
  </Row>
</Container>
```

* <i>The custom class demo-col is defined and used for the Styleguidist only.</i>