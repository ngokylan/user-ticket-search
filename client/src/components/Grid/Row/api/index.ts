import { getClass } from '../../../../_lib/helper';
import { RowResponsivePropertiesType } from '../type';

/**
 * Responsive properties
 */
export const responsiveProperties: RowResponsivePropertiesType[] = [
  { property: 'display', prefix: 'flex-' },
  { property: 'direction', prefix: 'flex-' },
  { property: 'wrap', prefix: 'flex-' },
  { property: 'justifyContent', prefix: 'justify-content-' },
  { property: 'alignItems', prefix: 'align-items-' },
  { property: 'alignContent', prefix: 'align-content-' },
];

/**
 *
 * @param block
 * @param props
 */
export function mapPropsToClass(block: string, props: any) {
  return getClass(
    block,
    {
      [`${props.className}`]: props.className,
      [`no-gutters`]: props.isNoGutters,
      [`flex-inline`]: props.isInline,
      [`flex-${props.direction}`]: props.direction,
      [`flex-${props.wrap}`]: props.wrap,
      [`justify-content-${props.justifyContent}`]: props.justifyContent,
      [`align-items-${props.alignItems}`]: props.alignItems,
      [`align-content-${props.alignContent}`]: props.alignContent,
      [`${mapResponsive(block, 'xs', props.xs)}`]: props.xs,
      [`${mapResponsive(block, 'sm', props.sm)}`]: props.sm,
      [`${mapResponsive(block, 'md', props.md)}`]: props.md,
      [`${mapResponsive(block, 'lg', props.lg)}`]: props.lg,
      [`${mapResponsive(block, 'xl', props.xl)}`]: props.xl,
      [`${mapResponsive(block, 'xxl', props.xxl)}`]: props.xxl,
    },
    []
  );
}

/**
 *
 * @param breakpoint
 * @param options
 */
export function mapResponsive(block: string, breakpoint: string, options: any) {
  if (options === null || options === undefined) {
    return '';
  }

  return responsiveProperties
    .map(item => {
      if (options[item.property] !== undefined) {
        return `${item.prefix}${breakpoint}-${options[item.property]}`;
      }
    })
    .join(' ');
}
