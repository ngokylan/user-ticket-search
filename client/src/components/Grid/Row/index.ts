export { default } from './Row';
export * from './type';
