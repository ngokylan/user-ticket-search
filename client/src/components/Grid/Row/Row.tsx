import React, { ReactNode } from 'react';
import { mapPropsToClass } from './api';
import { RowProps } from './type';
import './Row.scss';

function Row({ children, ...props }: RowProps): any {
  const className = mapPropsToClass('row', props);
  const { role, key } = props;

  return (
    <div className={className} key={key} role={role}>
      {children}
    </div>
  );
}

Row.defaultProps = {
  isNoGutters: false,
  isInline: false,
  role: 'row',
};

export default Row;
