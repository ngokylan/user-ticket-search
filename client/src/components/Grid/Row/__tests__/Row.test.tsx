import React from 'react';
import { mount } from 'enzyme';
import Row, { RowProps } from '..';

const defaultProps: RowProps = {
  isNoGutters: false,
  isInline: false,
  role: 'row',
};
const mountComponent = (customProps: any = {}) => {
  const props = { ...defaultProps, ...customProps };
  const wrapper = mount(<Row {...props} />);

  return { props, wrapper };
};

describe('Row', () => {
  describe('rendering', () => {
    it('should render component without any crash', () => {
      const { wrapper } = mountComponent();
      expect(wrapper).toHaveLength(1);
    });

    it('should be match with snapshot', () => {
      const { wrapper } = mountComponent();
      expect(wrapper).toMatchSnapshot();
    });

    it.each([
      ['isNoGutters', true],
      ['isNoGutters', false],
      ['isInline', true],
      ['isInline', false],
      ['direction', 'row'],
      ['direction', 'column'],
      ['direction', 'row-reverse'],
      ['direction', 'column-reverse'],
      ['wrap', 'wrap'],
      ['wrap', 'nowrap'],
      ['wrap', 'wrap-reverse'],
      ['justifyContent', 'start'],
      ['justifyContent', 'end'],
      ['justifyContent', 'center'],
      ['justifyContent', 'between'],
      ['justifyContent', 'around'],
      ['alignItems', 'start'],
      ['alignItems', 'end'],
      ['alignItems', 'center'],
      ['alignItems', 'baseline'],
      ['alignItems', 'stretch'],
      ['alignContent', 'start'],
      ['alignContent', 'end'],
      ['alignContent', 'center'],
      ['alignContent', 'between'],
      ['alignContent', 'around'],
      ['alignContent', 'stretch'],
      ['className', 'className'],
      ['className', null],
      ['className', undefined],
      ['className', false],
    ])(
      'should be match with snapshot when props %s is %s',
      (propsName, propsValue) => {
        const { wrapper } = mountComponent({ [propsName]: propsValue });
        expect(wrapper).toMatchSnapshot();
      }
    );
  });

  describe('props', () => {
    describe('default', () => {
      it('should have expected default props', () => {
        expect(JSON.stringify(defaultProps)).toBe(
          JSON.stringify(Row.defaultProps)
        );
      });
    });

    describe('children', () => {
      it('should render children', () => {
        const text = 'Hello world';
        const { wrapper } = mountComponent({ children: text });
        expect(wrapper.text()).toBe(text);
      });
    });
  });
});
