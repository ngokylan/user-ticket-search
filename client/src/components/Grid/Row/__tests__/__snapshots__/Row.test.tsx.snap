// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row rendering should be match with snapshot 1`] = `
<div
  className="row"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props alignContent is around 1`] = `
<div
  className="row align-content-around"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props alignContent is between 1`] = `
<div
  className="row align-content-between"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props alignContent is center 1`] = `
<div
  className="row align-content-center"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props alignContent is end 1`] = `
<div
  className="row align-content-end"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props alignContent is start 1`] = `
<div
  className="row align-content-start"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props alignContent is stretch 1`] = `
<div
  className="row align-content-stretch"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props alignItems is baseline 1`] = `
<div
  className="row align-items-baseline"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props alignItems is center 1`] = `
<div
  className="row align-items-center"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props alignItems is end 1`] = `
<div
  className="row align-items-end"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props alignItems is start 1`] = `
<div
  className="row align-items-start"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props alignItems is stretch 1`] = `
<div
  className="row align-items-stretch"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props className is className 1`] = `
<div
  className="row className"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props className is false 1`] = `
<div
  className="row"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props className is null 1`] = `
<div
  className="row"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props className is undefined 1`] = `
<div
  className="row"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props direction is column 1`] = `
<div
  className="row flex-column"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props direction is column-reverse 1`] = `
<div
  className="row flex-column-reverse"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props direction is row 1`] = `
<div
  className="row flex-row"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props direction is row-reverse 1`] = `
<div
  className="row flex-row-reverse"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props isInline is false 1`] = `
<div
  className="row"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props isInline is true 1`] = `
<div
  className="row flex-inline"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props isNoGutters is false 1`] = `
<div
  className="row"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props isNoGutters is true 1`] = `
<div
  className="row no-gutters"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props justifyContent is around 1`] = `
<div
  className="row justify-content-around"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props justifyContent is between 1`] = `
<div
  className="row justify-content-between"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props justifyContent is center 1`] = `
<div
  className="row justify-content-center"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props justifyContent is end 1`] = `
<div
  className="row justify-content-end"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props justifyContent is start 1`] = `
<div
  className="row justify-content-start"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props wrap is nowrap 1`] = `
<div
  className="row flex-nowrap"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props wrap is wrap 1`] = `
<div
  className="row flex-wrap"
  role="row"
/>
`;

exports[`Row rendering should be match with snapshot when props wrap is wrap-reverse 1`] = `
<div
  className="row flex-wrap-reverse"
  role="row"
/>
`;
